lang/cs/Org.Apache.REEF.Bridge.CLR/LocalObserver.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Tang.Implementations.InjectionPlan;
+using Org.Apache.REEF.Utilities.Logging;
+using Org.Apache.REEF.Wake.Avro;
+using Org.Apache.REEF.Wake.Remote;
+
+namespace Org.Apache.REEF.Bridge
+{
+    /// <summary>
+    /// The Local Observer class receives byte buffer messages from the transport layer,
+    /// deserializes the messages into Avro C# classes, and invokes the appropriate
+    /// IObserver callback on the Avro message observer.
+    /// </summary>
+    public sealed class LocalObserver : IObserver<IRemoteMessage<byte[]>>
+    {
+        /// <summary>
+        /// Specify the class that will be called to process deserialied Avro messages.
+        /// </summary>
+        [NamedParameter(Documentation =
+            "Must implement IObserver<IMessageInstance<TMessage>> for messages to be received.")]
+        public class MessageObserver : Name<object>
```

Comment:
How do you enforce it to implement  IObserver<IMessageInstance<TMessage>>? It is just an object in the form of NamedParameter.

Reply:
I wish we could figure it out. We use reflection on both C# and java side to auto-generate these structures (and pair them with the corresponding message handlers). I could not figure out how to use stricter types for such messages in Tang and Wake. If you can help us with that, let's run some experiments later and add type constraints in the subsequent PRs. We should file a JIRA issue for it, though.